packages/@lwc/engine/src/framework/stylesheet.ts

```diff
@@ -55,46 +66,54 @@ function insertGlobalStyle(styleContent: string) {
     }
 }
 
-function createStyleVNode(elm: HTMLStyleElement) {
-    const vnode = api.h(
-        'style',
-        {
-            key: 'style', // special key
-        },
-        EmptyArray
-    );
-    // TODO [#1364]: supporting the ability to inject a cloned StyleElement
-    // forcing the diffing algo to use the cloned style for native shadow
-    vnode.clonedElement = elm;
-    return vnode;
-}
-
 /**
  * Reset the styling token applied to the host element.
  */
-export function resetStyleAttributes(vm: VM): void {
-    const { context, elm } = vm;
+export function resetStyle(vm: VM): void {
+    const { context, elm, cmpRoot } = vm;
+    const { hostAttribute, styleElement } = context;
 
     // Remove the style attribute currently applied to the host element.
-    const oldHostAttribute = context.hostAttribute;
-    if (!isUndefined(oldHostAttribute)) {
-        removeAttribute.call(elm, oldHostAttribute);
+    if (!isUndefined(hostAttribute)) {
+        removeAttribute.call(elm, hostAttribute);
     }
 
     // Reset the scoping attributes associated to the context.
     context.hostAttribute = context.shadowAttribute = undefined;
+
+    // removing style tag if present
+    if (!isNull(styleElement) && !isUndefined(styleElement)) {
+        cmpRoot.removeChild(styleElement);
+        context.styleElement = null;
+    }
 }
 
 /**
  * Apply/Update the styling token applied to the host element.
  */
-export function applyStyleAttributes(vm: VM, hostAttribute: string, shadowAttribute: string): void {
-    const { context, elm } = vm;
+export function applyStyle(
+    vm: VM,
+    stylesheets: StylesheetFactory[],
+    stylesheetTokens: StylesheetTokens
+): void {
+    const { context, elm, cmpRoot } = vm;
+    const { hostAttribute, shadowAttribute } = stylesheetTokens;
     // Remove the style attribute currently applied to the host element.
     setAttribute.call(elm, hostAttribute, '');
 
     context.hostAttribute = hostAttribute;
     context.shadowAttribute = shadowAttribute;
+
+    // Caching style vnode so it can be removed when needed
+    const styleElm = (context.styleElement = evaluateCSS(
+        vm,
+        stylesheets,
+        hostAttribute,
+        shadowAttribute
+    ));
+    if (!isNull(styleElm)) {
+        cmpRoot.appendChild(styleElm);
```

Comment:
we should probably use insertBefore with null reference instead, since we never really use append at all, the diffing algo uses insertBefore, that minimize the number of dom APIs that we use.